### PR TITLE
feat: drain v3 issue and PR queue

### LINKED ISSUE #2154: [Bug]: Disabling IPV4 for agent nodes prevents cluster creation
### Description

Hi, I tried setting `disable_ipv4 = true` in my agent node pool, but this made terraform get stuck at connecting to the agent nodes. It would try to ssh to the node using `<ipv6_address::1>` address and keep on failing until timeout

### Kube.tf file

```terraform
module "kube-hetzner" {
  providers = {
    hcloud = hcloud
  }
  hcloud_token = var.hcloud_token

  # 1. For normal use, (the official version published on the Terraform Registry), use
  source = "kube-hetzner/kube-hetzner/hcloud"
  #    When using the terraform registry as source, you can optionally specify a version number.
  #    See https://registry.terraform.io/modules/kube-hetzner/kube-hetzner/hcloud for the available versions
  # version = "2.15.3"
  # 2. If you want to use the latest master branch (see https://developer.hashicorp.com/terraform/language/modules/sources#github), use
  # source = "github.com/kube-hetzner/terraform-hcloud-kube-hetzner"

  # Note that some values, notably "location" and "public_key" have no effect after initializing the cluster.
  # This is to keep Terraform from re-provisioning all nodes at once, which would lose data. If you want to update
  # those, you should instead change the value here and manually re-provision each node. Grep for "lifecycle".

  # * Your ssh public key
  ssh_public_key  = file("~/.ssh/id_ed25519.pub")
  ssh_private_key = file("~/.ssh/id_ed25519")

  # You can add additional SSH public Keys to grant other team members root access to your cluster nodes.
  # ssh_additional_public_keys = []

  # You can also add additional SSH public Keys which are saved in the hetzner cloud by a label.
  # See https://docs.hetzner.cloud/#label-selector
  # ssh_hcloud_key_label = "role=admin"

  # * For Hetzner locations see https://docs.hetzner.com/general/others/data-centers-and-connection/
  network_region = "us-east"

  # Using the default configuration you can only create a maximum of 42 agent-nodepools.

  # You can safely add or remove nodepools at the end of each list. That is due to how subnets and IPs get allocated (FILO).
  # The maximum number of nodepools you can create combined for both lists is 50 (see above).
  # Also, before decreasing the count of any nodepools to 0, it's essential to drain and cordon the nodes in question. Otherwise, it will leave your cluster in a bad state.

  # Before initializing the cluster, you can change all parameters and add or remove any nodepools. You need at least one nodepool of each kind, control plane, and agent.
  # ⚠️ The nodepool names are entirely arbitrary, but all lowercase, no special characters or underscore (dashes are allowed), and they must be unique.

  # Please note that changing labels and taints after the first run will have no effect. If needed, you can do that through Kubernetes directly.

  control_plane_nodepools = [
    {
      name        = "control-plane",
      server_type = "cpx21",
      location    = "ash",
      labels = [
        "node.kubernetes.io/role=control-plane"
      ],
      taints = [],
      count  = 3
      # swap_size   = "2G" # remember to add the suffix, examples: 512M, 1G
      # zram_size   = "2G" # remember to add the suffix, examples: 512M, 1G
      # kubelet_args = ["kube-reserved=cpu=250m,memory=1500Mi,ephemeral-storage=1Gi", "system-reserved=cpu=250m,memory=300Mi"]

      # Fine-grained control over placement groups (nodes in the same group are spread over different physical servers, 10 nodes per placement group max):
      placement_group = "control-planes"

      # Enable automatic backups via Hetzner (default: false)
      # backups = true

      # To disable public ips (default: false)
      # WARNING: If both values are set to "true", your server will only be accessible via a private network. Make sure you have followed
      # the instructions regarding this type of setup in README.md: "Use only private IPs in your cluster".
      # disable_ipv4 = true
      # disable_ipv6 = true
    }
  ]

  # I use these nodes to deploy smaller personal apps that don't need autoscaling + longhorn storage
  agent_nodepools = [
    {
      name        = "agent-sailboat",
      server_type = "cpx31",
      location    = "ash",
      labels = [
        "gport.giuliopime.dev/agent=sailboat",
        "node.longhorn.io/create-default-disk=true" # this allows longhord to use the memory for PV
      ],
      taints = [],
      count  = 2

      disable_ipv4 = trur
      # swap_size   = "2G" # remember to add the suffix, examples: 512M, 1G
      # zram_size   = "2G" # remember to add the suffix, examples: 512M, 1G
      # kubelet_args = ["kube-reserved=cpu=50m,memory=300Mi,ephemeral-storage=1Gi", "system-reserved=cpu=250m,memory=300Mi"]

      # Fine-grained control over placement groups (nodes in the same group are spread over different physical servers, 10 nodes per placement group max):
      # placement_group = "default"

      # Enable automatic backups via Hetzner (default: false)
      # backups = true

      # In the case of using Longhorn, you can use Hetzner volumes instead of using the node's own storage by specifying a value from 10 to 10240 (in GB)
      # It will create one volume per node in the nodepool, and configure Longhorn to use them.
      # Something worth noting is that Volume storage is slower than node storage, which is achieved by not mentioning longhorn_volume_size or setting it to 0.
      # So for something like DBs, you definitely want node storage, for other things like backups, volume storage is fine, and cheaper.

      # I use Hetzner Storage Boxes for slow storage instead of Volumes
      longhorn_volume_size = 0
    },
  ]

  # * LB location and type, the latter will depend on how much load you want it to handle, see https://www.hetzner.com/cloud/load-balancer
  load_balancer_type     = "lb11"
  load_balancer_location = "ash"

  # You can refine a base domain name to be use in this form of nodename.base_domain for setting the reverse dns inside Hetzner
  base_domain = "gport.giuliopime.dev"

  # Cluster Autoscaler
  # Providing at least one map for the array enables the cluster autoscaler feature, default is disabled.
  # ⚠️ Based on how the autoscaler works with this project, you can only choose either x86 instances or ARM server types for ALL autoscaler nodepools.
  # If you are curious, it's ok to have a multi-architecture cluster, as most underlying container images are multi-architecture too.

  # I use these nodes to deploy applications that need to scale automatically depending on the load
  # 'as' stands for auto-scaling
  autoscaler_nodepools = [
    {
      name        = "agent-as-sailboat"
      server_type = "cpx31"
      location    = "ash"
      min_nodes   = 3
      max_nodes   = 10
      labels = {
        "gport.giuliopime.dev/agent" : "as-sailboat"
      },
      taints = []
      # kubelet_args = ["kube-reserved=cpu=250m,memory=1500Mi,ephemeral-storage=1Gi", "system-reserved=cpu=250m,memory=300Mi"]
    }
    # {
    #   name        = "agent-racer"
    #   server_type = "cpx31"
    #   location    = "ash"
    #   min_nodes   = 0
    #   max_nodes   = 5
    #   labels      = {
    #     "gport.giuliopime.dev/agent": "racer"
    #   },
    #   taints      = [
    #     {
    #       key= "gport.giuliopime.dev/workload"
    #       value= "medium"
    #       effect= "NoExecute"
    #     }
    #   ]
    #   # kubelet_args = ["kube-reserved=cpu=250m,memory=1500Mi,ephemeral-storage=1Gi", "system-reserved=cpu=250m,memory=300Mi"]
    # },
  ]

  # To disable public ips on your autoscaled nodes, uncomment the following lines:
  autoscaler_disable_ipv4 = true
  # autoscaler_disable_ipv6 = true

  cluster_autoscaler_extra_args = [
    "--enforce-node-group-min-size=true",
  ]

  # To enable Hetzner Storage Box support, you can enable csi-driver-smb, default is "false".
  enable_csi_driver_smb = true

  # To use local storage on the nodes, you can enable Longhorn, default is "false".
  # See a full recap on how to configure agent nodepools for longhorn here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/373#discussioncomment-3983159
  # Also see Longhorn best practices here https://gist.github.com/ifeulner/d311b2868f6c00e649f33a72166c2e5b
  enable_longhorn = true

  # When you enable Longhorn, you can go with the default settings and just modify the above two variables OR you can add a longhorn_values variable
  # with all needed helm values, see towards the end of the file in the advanced section.
  # If that file is present, the system will use it during the deploy, if not it will use the default values with the two variable above that can be customized.
  # After the cluster is deployed, you can always use HelmChartConfig definition to tweak the configuration.

  # If you want to specify the Kured version, set it below - otherwise it'll use the latest version available.
  # See https://github.com/kubereboot/kured/releases for the available versions.
  kured_version = "1.17.1"

  ingress_controller = "nginx"

  # Use the klipperLB (similar to metalLB), instead of the default Hetzner one, that has an advantage of dropping the cost of the setup.
  # Automatically "true" in the case of single node cluster (as it does not make sense to use the Hetzner LB in that situation).
  # It can work with any ingress controller that you choose to deploy.
  # Please note that because the klipperLB points to all nodes, we automatically allow scheduling on the control plane when it is active.
  enable_klipper_metal_lb = "false"

  # By default nodes are drained before k3s upgrade, which will delete and transfer all pods to other nodes.
  # Set this to false to cordon nodes instead, which just prevents scheduling new pods on the node during upgrade
  # and keeps all pods running. This may be useful if you have pods which are known to be slow to start e.g.
  # because they have to mount volumes with many files which require to get the right security context applied.
  system_upgrade_use_drain = true

  # The cluster name, by default "k3s"
  cluster_name = "gport"

  # If you want to configure a different CNI for k3s, use this flag
  # possible values: flannel (Default), calico, and cilium
  # As for Cilium, we allow infinite configurations via helm values, please check the CNI section of the readme over at https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/#cni.
  # Also, see the cilium_values at towards the end of this file, in the advanced section.
  # ⚠️ Depending on your setup, sometimes you need your control-planes to have more than
  # 2GB of RAM if you are going to use Cilium, otherwise the pods will not start.
  cni_plugin = "calico"

  # Enables Hubble Observability to collect and visualize network traffic. Default: false
  # cilium_hubble_enabled = true

  # Configures the list of Hubble metrics to collect.
  # cilium_hubble_metrics_enabled = [
  #   "policy:sourceContext=app|workload-name|pod|reserved-identity;destinationContext=app|workload-name|pod|dns|reserved-identity;labelsContext=source_namespace,destination_namespace"
  # ]

  # IP Addresses to use for the DNS Servers, the defaults are the ones provided by Hetzner https://docs.hetzner.com/dns-console/dns/general/recursive-name-servers/.
  # The number of different DNS servers is limited to 3 by Kubernetes itself.
  # It's always a good idea to have at least 1 IPv4 and 1 IPv6 DNS server for robustness.
  dns_servers = [
    "1.1.1.1",
    "8.8.8.8",
    "2606:4700:4700::1111",
  ]

  # When this is enabled, rather than the first node, all external traffic will be routed via a control-plane loadbalancer, allowing for high availability.
  # The default is false.
  use_control_plane_lb = true

  # When the above use_control_plane_lb is enabled, you can change the lb type for it, the default is "lb11".
  control_plane_lb_type = "lb11"

  lb_hostname = "gport.giuliopime.dev"

  # It is best practice to turn this off, but for backwards compatibility it is set to "true" by default.
  # See https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/349
  # When "false". The kubeconfig file can instead be created by executing: "terraform output --raw kubeconfig > cluster_kubeconfig.yaml"
  # Always be careful to not commit this file!
  create_kubeconfig = false

  ### ADVANCED - Custom helm values for packages above (search _values if you want to located where those are mentioned upper in this file)
  # ⚠️ Inside the _values variable below are examples, up to you to find out the best helm values possible, we do not provide support for customized helm values.
  # Please understand that the indentation is very important, inside the EOTs, as those are proper yaml helm values.
  # We advise you to use the default values, and only change them if you know what you are doing!

  # You can inline the values here in heredoc-style (as the examples below with the <<EOT to EOT). Please note that the current indentation inside the EOT is important.
  # Or you can create a thepackage-values.yaml file with the content and use it here with the following syntax:
  # thepackage_values = file("thepackage-values.yaml")

  # Longhorn, all Longhorn helm values can be found at https://github.com/longhorn/longhorn/blob/master/chart/values.yaml
  # The following is an example, please note that the current indentation inside the EOT is important.
  #
  # autoscaler support is experimental, hoping we are good
  # defaultClass: false makes it so longhorn isn't used as the default storage class for all our deployed applications
  longhorn_values = <<EOT
defaultSettings:
  createDefaultDiskLabeledNodes: true
  kubernetesClusterAutoscalerEnabled: true
  defaultDataPath: /var/longhorn
  node-down-pod-deletion-policy: delete-both-statefulset-and-deployment-pod
persistence:
  defaultFsType: ext4
  defaultClassReplicaCount: 2
  defaultClass: false
  EOT
}
```

### Screenshots

_No response_

### Platform

Mac

### LINKED ISSUE #2173: [Bug]: Load balancer fails to lookup "custom" on staging branch when using custom ingress
### Description:
I was playing around with the staging branch because I wanted to deploy my own ingress into the cluster. When setting it up I ran into some issues though, since the init script tried to lookup values, that did not exist or ran into timeouts during the deployment. The latter one is probably an unrelated SSH issue, but I'm adding it here anyway.

### Errors:
```bash
module.kube-hetzner.terraform_data.agents["0-2-worker-fsn-001"]: Still creating... [04m18s elapsed]
module.kube-hetzner.terraform_data.agents["0-2-worker-fsn-001"]: Still creating... [04m25s elapsed]
╷
│ Error: remote-exec provisioner error
│ 
│   with module.kube-hetzner.terraform_data.agents["0-1-worker-fsn-001"],
│   on ../../terraform-hcloud-kube-hetzner/agents.tf line 274, in resource "terraform_data" "agents":
│  274:   provisioner "remote-exec" {
│ 
│ error executing "/tmp/terraform_291093008.sh": Process exited with status 124
╵
╷
│ Error: remote-exec provisioner error
│ 
│   with module.kube-hetzner.terraform_data.agents["0-2-worker-fsn-001"],
│   on ../../terraform-hcloud-kube-hetzner/agents.tf line 274, in resource "terraform_data" "agents":
│  274:   provisioner "remote-exec" {
│ 
│ error executing "/tmp/terraform_1518748349.sh": Process exited with status 124
╵
╷
│ Error: Invalid function argument
│ 
│   on ../../terraform-hcloud-kube-hetzner/init.tf line 741, in resource "terraform_data" "kustomization":
│  741:       until [ -n "\$(kubectl get -n ${local.ingress_controller_namespace} service/${lookup(local.ingress_controller_service_names, var.ingress_controller)} --output=jsonpath='{.status.loadBalancer.ingress[0].${var.lb_hostname != "" ? "hostname" : "ip"}}' 2> /dev/null)" ]; do
│     ├────────────────
│     │ while calling lookup(inputMap, key, default...)
│     │ local.ingress_controller_service_names is object with 3 attributes
│ 
│ Invalid value for "inputMap" parameter: the given object has no attribute "custom".
```

```bash
module.kube-hetzner.module.agents["0-2-worker-fsn-001"].hcloud_server.server (remote-exec):   Target Platform: unix
╷
│ Error: remote-exec provisioner error
│ 
│   with module.kube-hetzner.module.agents["0-2-worker-fsn-001"].hcloud_server.server,
│   on ../../terraform-hcloud-kube-hetzner/modules/host/main.tf line 86, in resource "hcloud_server" "server":
│   86:   provisioner "remote-exec" {
│ 
│ timeout - last error: dial tcp XXX.XXX.XXX.XXX:2222: connect: connection refused
```

### Possible Solution:
I adjusted https://github.com/mysticaltech/terraform-hcloud-kube-hetzner/blob/edfd1cf2d2699fb5e81c62f6e5ba3bc314921be5/locals.tf#L866 to the following

```tf
skip_ingress_lb_wait       = local.has_external_load_balancer_base || var.ingress_controller == "custom"
```

to fix it. Somewhere in the multiple merge requests the parameter for skipping the load balancer logic was changed from `has_external_load_balancer`, which included a special handling for "custom", to `skip_ingress_lb_wait`, if I remember correctly.

### Kube.tf file

```terraform
terraform {
  required_providers {
    hcloud = {
      source  = "hetznercloud/hcloud"
      version = "~> 1.45"
    }
  }
}

provider "hcloud" {
  token = var.hcloud_token
}

module "kube-hetzner" {
  source = "../../terraform-hcloud-kube-hetzner/"

  providers = {
    hcloud = hcloud
  }

  hcloud_token = var.hcloud_token

  ssh_public_key  = file("~/.ssh/id_ed25519.pub")
  ssh_private_key = file("~/.ssh/id_ed25519")

  network_region = "eu-central" # change to `us-east` if location is ash

  ssh_port = 2222

  ssh_max_auth_tries = 10

  control_plane_nodepools = [
    {
      name         = "cp-fsn-001"
      server_type  = "cx23"
      location     = "fsn1"
      os           = "microos"
      kubelet_args = ["kube-reserved=cpu=250m,memory=1500Mi,ephemeral-storage=1Gi", "system-reserved=cpu=250m,memory=300Mi"]

      nodes = {
        "1" : {}
      }

      labels = []
      taints = []

      disable_ipv6 = true

    },
  ]

  agent_nodepools = [
    {
      name         = "worker-fsn-001"
      server_type  = "cx33"
      location     = "fsn1"
      os           = "microos"
      kubelet_args = ["kube-reserved=cpu=50m,memory=300Mi,ephemeral-storage=1Gi", "system-reserved=cpu=250m,memory=300Mi"]

      nodes = {
        "1" : {},
        "2" : {}
      }

      labels = [
        "node.kubernetes.io/role=worker",
        "node.longhorn.io/create-default-disk=config",
        "node.kubernetes.io/server-usage=storage"
      ]
      taints = []

      disable_ipv6 = true

      longhorn_volume_size = 10
      longhorn_mount_path = "/var/longhorn"
    },
  ]

  load_balancer_type     = "lb11"
  load_balancer_location = "fsn1"

  load_balancer_disable_ipv6 = true

  k3s_encryption_at_rest = true

  enable_longhorn = true

  ingress_controller = "custom"

  automatically_upgrade_os = false

  firewall_kube_api_source = ["myipv4"]

  firewall_ssh_source = ["myipv4"]

  extra_firewall_rules = [
    {
      description     = "Allow ArgoCD access to resources via SSH"
      direction       = "out"
      protocol        = "tcp"
      port            = "22"
      source_ips      = [] # Won't be used for this rule
      destination_ips = ["0.0.0.0/0", "::/0"]
    }
  ]

  enable_x86 = true
  enable_arm = false

  enable_cert_manager = false

  longhorn_merge_values = <<-EOT
defaultSettings:
  createDefaultDiskLabeledNodes: true
  kubernetesClusterAutoscalerEnabled: false # Notify Longhorn if the autoscaler is active.
  defaultDataPath: /var/lib/longhorn
  defaultReplicaCount: 3
  nodeDownPodDeletionPolicy: delete-both-statefulset-and-deployment-pod # Force delete any pods that are stuck on nodes that are down.
persistence:
  defaultClass: true
  defaultFsType: ext4
  defaultClassReplicaCount: 3
  EOT

  user_kustomizations = {
    "1" = {
      source_folder        = "longhorn-storage-classes"
      kustomize_parameters = {}
      pre_commands         = ""
      post_commands = "" #"kubectl annotate nodes -l node.kubernetes.io/server-usage=storage node.longhorn.io/default-disks-config='[ { \"path\":\"/var/lib/longhorn\",\"allowScheduling\":true, \"storageReserved\":21474836480, \"tags\":[ \"nvme\" ]}, { \"name\":\"hcloud-volume\", \"path\":\"/var/longhorn\",\"allowScheduling\":true, \"storageReserved\":1073741824,\"tags\":[ \"ssd\" ] } ]' --overwrite"
    }
  }
}
```

### Screenshots

_No response_

### Platform

Linux

### LINKED ISSUE #2178: Cilium: smarter defaults for DSR, Maglev, and geneve dispatch
## Context

The module knows enough about the deployment environment to set better Cilium defaults than it currently does. Specifically, it knows:

- It's running on Hetzner Cloud (not bare metal, not another cloud)
- Whether WireGuard is enabled (`enable_wireguard`)
- Whether kube-proxy is disabled (`disable_kube_proxy`)
- The CNI is Cilium (`cni_plugin = "cilium"`)
- The routing mode (`cilium_routing_mode`)

Yet the defaults leave several Cilium performance features off, requiring users to discover and configure them via `cilium_merge_values`.

## Current behavior

| Setting | Current default | Effect |
|---------|----------------|--------|
| `loadBalancer.mode` | `snat` (Cilium default) | Reply packets go through the LB node, adding latency |
| `loadBalancer.algorithm` | `random` (Cilium default) | No consistent hashing, connections shuffle on node changes |
| `loadBalancer.dsrDispatch` | `opt` (Cilium default) | IP options — not encrypted by WireGuard, may be dropped by cloud fabrics |
| `kubeProxyReplacement` | `true` (hardcoded) | Always set, even when `disable_kube_proxy = false` — creates redundant iptables |

### When `disable_kube_proxy = true` and `cni_plugin = "cilium"`

Enable DSR + Maglev, since the module already hardcodes `kubeProxyReplacement: true`:

```yaml
loadBalancer:
  mode: dsr
  algorithm: maglev
```

### When `enable_wireguard = true`

Force geneve dispatch instead of the default IP options. With `opt` dispatch, the source IP of DSR-forwarded packets is the external client's IP (not a known WireGuard peer), so WireGuard cannot encrypt them (see cilium/cilium#32180). This also requires `tunnelProtocol: geneve`:

```yaml
loadBalancer:
  dsrDispatch: geneve
tunnelProtocol: geneve
```

This is safe on Hetzner Cloud because inter-node traffic goes over Cloud Networks (private), which are not filtered by the Hetzner Cloud Firewall ([Firewall FAQ](https://docs.hetzner.com/cloud/firewalls/faq/): "Can Firewalls secure traffic to my private Hetzner Cloud Networks? Not yet") — so UDP 6081 (geneve) flows freely.

### When `disable_kube_proxy = false`

Either don't set `kubeProxyReplacement: true` (let Cilium auto-detect), or warn users that both kube-proxy and Cilium's replacement are running simultaneously. Currently the module always hardcodes `kubeProxyReplacement: true` regardless of this setting (line 624 in `locals.tf`, outside any `%{if}` block).

## Performance impact

Published benchmarks for each feature:

**DSR vs SNAT:** P99 latency ~30-40% reduction — reply packets skip the LB node entirely. Return path latency cut ~50%. Sources: Cilium tuning guides, [Fractured Architecture 10G blog](https://www.fracturedarchitecture.com/architecture-the-10g-ebpf-edge-with-opnsense-istio-ready-cilium-dsr/).

**Maglev vs random:** ~1% connection disruption on backend topology change vs full reassignment. O(1) lookup, same as random — no steady-state difference. Matters for long-lived TCP/WebSockets/gRPC during rolling updates. Source: [Google Maglev paper (NSDI 2016)](https://research.google.com/pubs/archive/44824.pdf).

**kube-proxy disabled (eBPF) vs iptables:** O(1) hash map vs O(n) rule chain walk. At 10k services, iptables uses ~80k rules, ~500MB RAM, ~35% CPU core. Real-world: [Trendyol migration](https://medium.com/trendyol-tech/trendyols-cilium-transformation-ff064ff79fce) — 40% throughput increase, 12% packets/sec improvement.

**XDP** (already the module default via `best-effort`): [Cilium 1.8 benchmarks](https://cilium.io/blog/2020/06/22/cilium-18/) — 2.8x throughput vs TC, 4.3x vs iptables. Hetzner Cloud VMs use virtio-net which supports XDP natively.

## References

- [Cilium DSR mode docs](https://docs.cilium.io/en/v1.18/network/kubernetes/kubeproxy-free/#dsr-mode)
- [Cilium Maglev docs](https://docs.cilium.io/en/v1.18/network/kubernetes/kubeproxy-free/#maglev-consistent-hashing)
- cilium/cilium#32180 — WireGuard cannot encrypt DSR opt-dispatch packets
- cilium/cilium#30845 — DSR geneve + native routing requires `tunnelProtocol: geneve`
- [Hetzner Cloud Firewalls FAQ](https://docs.hetzner.com/cloud/firewalls/faq/)
- [Azure CNI eBPF benchmarks (Microsoft)](https://techcommunity.microsoft.com/blog/azurenetworkingblog/introducing-ebpf-host-routing-high-performance-ai-networking-with-azure-cni-powe/4468216)

### PR DESCRIPTION
**Pull request by mysticaltech**
## Summary

This is the v3 queue-drain umbrella PR. The goal is to bring the open issue and PR queue to zero, apart from the standing v3 promotion PR, while preserving contributor history for useful PRs and refusing changes that push kube-hetzner in the wrong direction.

- Integrates vetted staging PRs while preserving original contributor commits/history: #2182, #2166, #2165, #2160, and #2158.
- Fixes #2154 and #2173 directly on the v3 branch.
- Addresses #2178 by aligning v3 Cilium/CCM defaults with dual-stack, while avoiding the risky subnet rewrite from #2170.
- Documents the #2179 Longhorn/autoscaler boundary instead of shipping unsafe orphan-volume behavior.
- Carries forward the already-merged v2.19.3 cleanup from `master` into `staging` via #2183.
- Leaves #2147 open intentionally as the separate v3 staging-to-master promotion PR.

## Queue State

- Open issues: 0.
- Open PRs after this intake: #2185 and #2147 only.
- #2147 is v3 release promotion work, not an untriaged contribution PR.

## Active v3 Intake Decisions

| Item | Decision | Rationale |
| --- | --- | --- |
| #2182 | Accepted with contributor history preserved | Useful v3 fix for private NIC rename handling; hardened with a fixup to refresh stale eth1 udev rules when the interface is already renamed. |
| #2166 | Accepted with contributor history preserved | Useful NAT router/private bastion improvement; hardened so NAT router provisioners honor private bastion mode. |
| #2165 | Accepted with contributor history preserved | Useful NAT router `extra_runcmd` support; hardened so changes rerun through Terraform instead of being creation-only cloud-init. |
| #2160 | History preserved, implementation replaced | Original intent was valid, but the diff was stale. Replaced with a v3-safe user-kustomization replacement trigger and fixed upload/deploy edge cases. |
| #2158 | Accepted with contributor history preserved | Useful per-nodepool/node `os_snapshot_id` support; hardened null/empty override handling and avoided unused snapshot lookups. |
| #2170 | Not merged as-is; safe pieces salvaged | The subnet rewrite was too high-risk and not aligned with this v3 intake. Only safe dual-stack Cilium/CCM rendering/address-family pieces were carried forward. |
| #2154 | Fixed directly | Autoscaler default route logic now separates private IPv4, public IPv4, and public IPv6 paths. |
| #2173 | Fixed directly | Custom ingress now skips the managed LB wait path. |
| #2178 | Fixed conservatively | v3 dual-stack Cilium/CCM defaults are improved, but automatic DSR/Maglev defaults were not enabled because they are topology-sensitive and unsafe as universal defaults. Users can still opt in via Cilium values. |
| #2179 | Closed as docs/support boundary | Autoscaler-managed Longhorn volumes would require write-capable Hetzner token material in user-data and can orphan volumes on scale-down. Longhorn remains static-nodepool guidance. |
| #2177 | Closed/deferred | SSH through the control-plane LB is architectural. The current LB is the Kubernetes API path, not a general SSH transport. |
| #2169 | Closed as upstream behavior | Longhorn NFS behavior belongs upstream/user configuration; no risky module behavior change shipped. |

## Already Landed Through v2.19.3 And Included In Staging

These were part of the earlier patch train, have already shipped through #2183, and are now included in `staging` before this v3 queue-drain PR:

| Item | Decision | Rationale |
| --- | --- | --- |
| #2153 / #2152 | Accepted and fixed | NAT router null validation was real and patch-safe. The implementation was extended beyond the narrow PR so all optional/null `nat_router` validation paths are safe. |
| #2162 / #2161 | Accepted and fixed | Autoscaler zram cloud-init deadlock was real and patch-safe. Removed the bad ordering that could deadlock autoscaler bootstrap. |
| #2163 | Accepted and fixed | Debian 12 NAT router fail2ban behavior was real. Hardened package/backend/journal matching and service start/restart behavior. |
| #2167 | Accepted and fixed | MicroOS snapper timeline growth was a real operational footgun. Reduced retention safely without changing user-facing module shape. |
| #2174 / #2180 | Accepted as part of a fuller fix | The submitted resize trigger was useful but incomplete alone. v2.19.3 fixed Longhorn fstab UUIDs, stale fstab entries, resize triggers, and ext4/xfs resize behavior together. |
| #2172 | Fixed directly | System-upgrade Plan trigger inputs now include drain/eviction controls so post-provision changes apply. |
| #2176 | Fixed directly | Control-plane LB now has an explicit HTTPS `/readyz` health check on port 6443 while keeping service protocol TCP. |
| #2168 | Closed as docs/support | Existing `hetzner_csi_values = file(...)` support was documented instead of adding duplicate module surface. |
| #2155 | Fixed directly | Removed the child-module `github` provider block and replaced release lookup with unauthenticated `hashicorp/http`, keeping callers compatible with `count`, `for_each`, and `depends_on`. |

## Strategic Guardrails

- Preserve contributor attribution for accepted PRs: contributor commits are kept in history, with Karim-owned fixup commits added where needed.
- Do not accept PRs just because they exist: stale, risky, or misaligned diffs are either replaced with a safer implementation or closed with rationale.
- Keep v3 direction coherent: dual-stack improvements belong here, but topology-sensitive behavior like DSR/Maglev stays opt-in instead of becoming a surprise default.
- Keep Longhorn/autoscaler behavior honest: do not create automation that can leak credentials, orphan volumes, or imply lifecycle guarantees the module cannot safely provide.
- Use maintainer judgment for project-shaping calls: when an item is architectural rather than a bounded fix, close/defer with rationale instead of smuggling it into the queue drain.

## Review

- Gemini full-diff review: ALL CLEAR before targeted hardening.
- Codex CLI review used `gpt-5.5` with `model_reasoning_effort=xhigh` and found Cilium DSR, HCCM route CIDR, legacy CCM toleration, NAT router update-path, and user-kustomization edge cases.
- The current head removes automatic DSR/Maglev defaults, keeps HCCM routes IPv4-only, sequences NAT router provisioners safely, and fixes user-kustomization upload/deploy triggers.

## Tests

- `terraform fmt -recursive`
- `terraform-docs .`
- `terraform validate -no-color`
- `git diff --check`
- In `/Volumes/MysticalTech/Code/kube-test`: `terraform init -upgrade && terraform plan -no-color`
  - Result: `Plan: 39 to add, 0 to change, 0 to destroy` on the fresh `test12` test config.

Closes #2154
Closes #2173
Closes #2178

### CODE REVIEW

File: nat-router.tf [hunk 1]
```diff
@@ -232,7 +255,7 @@ resource "terraform_data" "nat_router_fail2ban" {
   count = var.nat_router != null ? (var.nat_router.enable_redundancy ? 2 : 1) : 0
 
   depends_on = [
-    terraform_data.nat_router_await_cloud_init,
+    terraform_data.nat_router_fail2ban,
```
[the remark this answers is not in the document]
Comment (mysticaltech) [reply nat-router.tf]: Resolved. The self-reference was fixed in df3bd9e so nat_router_fail2ban depends on nat_router_await_cloud_init again. I also pushed bf2e690 to harden the sequencing: control planes and agents now wait for nat_router_fail2ban and nat_router_extra_runcmd before node provisioning.

[the remark this answers is not in the document]
Comment (mysticaltech) [reply nat-router.tf]: Resolved. The self-reference was fixed in df3bd9e so nat_router_fail2ban depends on nat_router_await_cloud_init again. I also pushed bf2e690 to harden the sequencing: control planes and agents now wait for nat_router_fail2ban and nat_router_extra_runcmd before node provisioning.

**Comment by mysticaltech** [comment]
Final review pass pushed in bf2e690. Addressed the automated NAT fail2ban dependency review, added node sequencing after NAT hardening/custom commands, made Cilium Egress Gateway validate its kube-proxy replacement requirement, and fixed the HA reconciler to handle cluster-scoped CiliumEgressGatewayPolicy resources. Verification: terraform fmt -recursive, terraform validate -no-color, git diff --check, Gemini focused review = ALL CLEAR, kube-test terraform init -upgrade + plan = 39 to add, 0 to change, 0 to destroy, replace=0.
